src/audio/drc/drc_generic.c

```diff
@@ -0,0 +1,435 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Google LLC. All rights reserved.
+//
+// Author: Pin-chih Lin <johnylin@google.com>
+
+#include <sof/audio/component.h>
+#include <sof/audio/drc/drc.h>
+#include <sof/audio/drc/drc_math.h>
+#include <sof/audio/format.h>
+#include <stdint.h>
+
+/* This is the knee part of the compression curve. Returns the output level
+ * given the input level x. */
+static float knee_curveK(struct sof_drc_params *p, float x)
+{
+	const float knee_alpha = Q_CONVERT_QTOF(p->knee_alpha, 20);
+	const float knee_beta = Q_CONVERT_QTOF(p->knee_beta, 30);
+	const float K = Q_CONVERT_QTOF(p->K, 12);
+
+	/* The formula in knee_curveK is linear_threshold +
+	 * (1 - expf(-k * (x - linear_threshold))) / k
+	 * which simplifies to (alpha + beta * expf(gamma))
+	 * where alpha = linear_threshold + 1 / k
+	 *	 beta = -expf(k * linear_threshold) / k
+	 *	 gamma = -k * x
+	 */
+	return knee_alpha + knee_beta * knee_expf(-K * x);
+}
+
+/* Full compression curve with constant ratio after knee. Returns the ratio of
+ * output and input signal. */
+static float volume_gain(struct sof_drc_params *p, float x)
+{
+	const float knee_threshold = Q_CONVERT_QTOF(p->knee_threshold, 20);
+	const float linear_threshold = Q_CONVERT_QTOF(p->linear_threshold, 20);
+	const float ratio_base = Q_CONVERT_QTOF(p->ratio_base, 30);
+	const float slope = Q_CONVERT_QTOF(p->slope, 20);
+	float y;
+
+	if (x < knee_threshold) {
+		if (x < linear_threshold)
+			return 1;
+		y = knee_curveK(p, x) / x;
+	} else {
+		/* Constant ratio after knee.
+		 * log(y/y0) = s * log(x/x0)
+		 * => y = y0 * (x/x0)^s
+		 * => y = [y0 * (1/x0)^s] * x^s
+		 * => y = ratio_base * x^s
+		 * => y/x = ratio_base * x^(s - 1)
+		 * => y/x = ratio_base * e^(log(x) * (s - 1))
+		 */
+		y = ratio_base * knee_expf(logf(x) * (slope - 1));
```

Comment:
I'm guessing this is already in your list of things to do but probably good idea to write own fractional version of logf() to avoid the C float math library.